Release/snowplow fractribution/0.1.0

agnes-kiss opened the pull request:
## Description & motivation
Add support for Snowflake


## Checklist
- [x] I have verified that these changes work locally
- [x] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [x] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link here if required)

## Release Only Checklist
- [x] I have updated the version number in all relevant places
- [x] I have update the Github pages `dbt docs` 
- [x] I have updated the CHANGELOG.md

zloff committed:
Add support for Snowflake

agnes-kiss committed:
Prepare for release

rlh1994 commented on models/snowflake/snowplow_fractribution_conversions_by_customer_id.sql:
```diff
@@ -0,0 +1,36 @@
+WHERE
+    {{ conversion_clause() }}
+    AND
+    DATE(derived_tstamp) >= CASE WHEN '{{ var('conversion_window_start_date') }}' = '' 
+                                THEN current_date()-31
+                                ELSE '{{ var('conversion_window_start_date') }}'
+                                END
+    AND
+    DATE(derived_tstamp) <= CASE WHEN '{{ var('conversion_window_end_date') }}' = '' 
+                                THEN current_date()-1
+                                ELSE '{{ var('conversion_window_end_date') }}'
+                                END
```
This worries me a little as it has the potential to hit the issue we've seen before about not pruning queries correctly. It's probably fine because it's a small table(?) but maybe something to keep in mind.

rlh1994 commented on models/snowflake/snowplow_fractribution_path_summary.sql:
```diff
@@ -0,0 +1,26 @@
+    GROUP BY transformedPath
+)
+SELECT
+  IFNULL(PathsToConversion.transformedPath,
+         PathsToNonConversion.transformedPath) AS transformedPath,
+  IFNULL(PathsToConversion.conversions, 0) AS conversions,
+  IFNULL(PathsToNonConversion.nonConversions, 0) AS nonConversions,
+  PathsToConversion.revenue
+FROM PathsToConversion
+FULL JOIN PathsToNonConversion
+  USING(transformedPath)
```
I think coaleasce is more db-independent, and I don't particularly like the `USING` clause, but not a blocker for this release.

rlh1994 commented on models/snowflake/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -0,0 +1,32 @@
+  con.customerId,
+  conversionTimestamp,
+  revenue,
+  ARRAY_TO_STRING({{schema}}.TrimLongPath(
+    ARRAY_AGG(channel) WITHIN GROUP (ORDER BY visitStartTimestamp), {{ var('path_lookback_steps') }}),
+    ' > ') AS path,
+  ARRAY_TO_STRING(
+    {% for path_transform_name, _ in var('path_transforms')|reverse %}
+      {{schema}}.{{path_transform_name}}(
+    {% endfor %}
+        ARRAY_AGG(channel) WITHIN GROUP (ORDER BY visitStartTimestamp)
+    {% for _, arg_str in var('path_transforms') %}
+      {% if arg_str %}, {{arg_str}}{% endif %})
+    {% endfor %}
+    , ' > ') AS transformedPath
```
Again for future multi-db support, could these be replaced by https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#listagg

rlh1994 commented on models/snowflake/snowplow_fractribution_sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,56 @@
+        page_views.domain_userid = user_mapping.domain_userid
+    {% endif %}
+WHERE
+    DATE(derived_tstamp) >= DATEADD(d, -{{ var('path_lookback_days') + 1 }}, 
+                                CASE WHEN '{{ var('conversion_window_start_date') }}' = '' 
+                                THEN current_date()-31
+                                ELSE '{{ var('conversion_window_start_date') }}'
+                                END)
+    AND
+    DATE(derived_tstamp) <= CASE WHEN '{{ var('conversion_window_end_date') }}' = '' 
+                                THEN current_date()-1
+                                ELSE '{{ var('conversion_window_end_date') }}'
+                                END
+    AND
```
Again just a little worry about optimisation long term here.

rlh1994 commented on models/snowflake/snowplow_fractribution_sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,56 @@
+                                END
+    AND
+    -- restrict to certain hostnames
+    page_urlhost IN (
+        {%- for urlhost in var('conversion_hosts') %}
+            '{{ urlhost }}'
+            {%- if not loop.last %},{% endif %}
+        {%- endfor %}
+    )
```
If this variable is null or an empty list this will break the query - will this always have a value and you'd never want to just include all of them?

agnes-kiss replied:
It should be set and there is a default string they would have to overwrite with '' or null but nothing is stopping them so  I have added some exception handling just to get a proper error message if they do that.

rlh1994 reviewed (approved):
A few comments but nothing blocking and things that can be addressed later. 
1. It would be good to document the macros at some point as currently there is no `.yml` file for them.
2. Mostly an FYI, the release workflow only pulls text if there is a line with the old version (default 0.0.1), so you'll need to edit the release text yourself once it's there.

agnes-kiss commented:
Thanks for looking more into the overall code as well, Ryan! I have added a yml for the macros and added exception handling for the potentially missing variable, then replaced the dbt docs as well. For the remaining topics I think we can wait until the next release as it will most likely be about adapter support changes (there will be even more things to harmonise in this regard) and also more time to think them through.

rlh1994 reviewed (approved):
LGTM!
